src/Features/LanguageServer/Protocol/Handler/Diagnostics/DiagnosticSources/WorkspaceDocumentDiagnosticSource.cs

```diff
@@ -43,7 +29,8 @@ internal sealed record WorkspaceDocumentDiagnosticSource : AbstractDocumentDiagn
             // We call GetDiagnosticsForIdsAsync as we want to ensure we get the full set of diagnostics for this document
             // including those reported as a compilation end diagnostic.  These are not included in document pull (uses GetDiagnosticsForSpan) due to cost.
             // However we can include them as a part of workspace pull when FSA is on.
-            var documentDiagnostics = await diagnosticAnalyzerService.GetDiagnosticsForIdsAsync(Document.Project.Solution, Document.Project.Id, Document.Id, cancellationToken: cancellationToken).ConfigureAwait(false);
+            var documentDiagnostics = await diagnosticAnalyzerService.GetDiagnosticsForIdsAsync(
+                Document.Project.Solution, Document.Project.Id, Document.Id, cancellationToken: cancellationToken).ConfigureAwait(false);
```

Comment:
this part was a little messy.  specifically, GetDiagnosticsForIdsAsync doesn't support fine-grained diagnostic requests (@mavasani ).  So for workspace-document-pull requests, we just wait on getting all the diagnostics.  this means that there's no speedup on getting diagnostics for closed files.  However, i think that's ok as you only get those for FSA, and FSA is notoriously slow anyways.

If we *do* want to support this, it would be possible. We'd just need GetDiagnosticsForIdsAsync to support passing in a diagnostickind.  I initially attempted to do that, but i was uncertain about the precise semantics of how that would work in the impl.

If you are interested in adding support for that @mavasani  i can update this side of things.

Reply:
I think it is fine to wait for adding this support as it does not seem to be a mainline scenario.